### PR TITLE
support empty batch in group normalization

### PR DESCRIPTION
**Pull request by ppwwyyxx**
Summary: https://github.com/pytorch/pytorch/issues/12013

Differential Revision: D19463720

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 7c75be24:


**None of the build failures appear to be your fault**.

* **1/1** [broken upstream](https://dr.pytorch.org/master-timeline.html?count=1&sha1=58234c0254f135d1e3714b17f44f4922d56d41b4&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35) at merge base 58234c02 since Jan 18
  <details>
  <summary>
  Please rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch origin viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch origin viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=7c75be24975c42901fe318ecf86cbc86c452e402).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably **caused by upstream breakages**:
* [pytorch_linux_backward_compatibility_check_test](https://dr.pytorch.org/build-details.html?build_id=88362912) from Jan 18



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: test/test_nn.py [hunk 1]
```diff
@@ -8919,8 +8919,11 @@ def _test_module_empty_input(self, module, inp):
         out.backward(gO)
         self.assertEqual(out.size(), inp.size())
         for p in module.parameters():
+            # TODO: p.grad should not be None, but this is not yet supported
+            # (https://github.com/pytorch/pytorch/issues/12013)
```
Comment (SsnL) [inline test/test_nn.py]: Which part of the linked issue talks about `None` grad? I couldn't find it.

**Review by ssnl — APPROVED** [review]
lgtm

File: aten/src/ATen/native/Normalization.cpp [hunk 2]
```diff
@@ -593,7 +593,8 @@ Tensor group_norm(const Tensor& input, int64_t num_groups,
              " and input of shape ", input.sizes());
 
     // Apply group norm
-    auto input_reshaped = input.contiguous().view({1, b * num_groups, -1});
+    // view(..., -1) does not work for empty tensor
+    auto input_reshaped = input.contiguous().view({1, b * num_groups, b ? -1 : 1});
```
Comment (vadimkantorov) [inline aten/src/ATen/native/Normalization.cpp]: should we also file an issue about `view` not supporting -1 for empty tensors?

Comment (SsnL) [reply aten/src/ATen/native/Normalization.cpp]: It is deliberately not supported due to ambiguity.

File: test/test_nn.py (hunk 1 above)
Comment (ppwwyyxx) [reply test/test_nn.py]: https://github.com/pytorch/pytorch/issues/12013#issuecomment-575871264 says `p.grad` is None

File: aten/src/ATen/native/Normalization.cpp (hunk 2 above)
Comment (ppwwyyxx) [reply aten/src/ATen/native/Normalization.cpp]: TensorFlow does support reshape empty tensors using -1, by just ignoring the empty dims. It actually works quite well in practice (though not so intuitively when facing wrong inputs). For example this current line I'm changing would work out of the box.
